ACME: use Cryptography (if a new enough version is available) instead of OpenSSL

felixfontein opened the pull request:
##### SUMMARY
Advantages:
 - no need for `openssl` binary to be available;
 - no more hackish "pass this data to `openssl`, and then parse its output", which can break with new OpenSSL releases;
 - private keys don't need to be written to disk to be of use anymore (well, Ansible still does that when passing the module bundled with the arguments to the target machine -- but we don't do it explicitly again) (see #32915, #32948, #34979).

Implements and closes #36363.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
acme_account
acme_certificate
acme_certificate_revoke

##### ANSIBLE VERSION
```
2.7.0
```

##### ADDITIONAL INFORMATION
Tested against Pebble and both Let's Encrypt staging endpoints.

TODO:
 - ~~update documentation;~~
 - ~~actually test `account_key_content` use (the current tests in #41626 don't do that yet);~~
 - ~~add option to force usage of OpenSSL (even if new enough `cryptography` module is there);~~
 - determine which version of `cryptography` is actually required (I guessed 2.0, but that must not be true).

felixfontein committed:
Collecting PEM -> DER conversions.

felixfontein committed:
Using cryptography instead of OpenSSL binary in some situations.

felixfontein committed:
Moving key-to-disk writing for key content to parse_account_key.

felixfontein committed:
Rename parse_account_key -> parse_key.

felixfontein committed:
Move OpenSSL specific code for key parsing and request signing into global functions.

felixfontein committed:
Also using cryptography for key parsing and request signing.

felixfontein committed:
Remove assert statements.

felixfontein committed:
Fixing handling of key contents for cryptography code path.

felixfontein committed:
Allow to disable the use of cryptography.

felixfontein committed:
Updating documentation.

felixfontein commented:
I managed to install cryptography==2.0 on my machine, and it worked fine (with the Pebble tests). I couldn't install an older version (there were always build errors for the libssl integration), so no idea whether some of them will work.
I'll probably create a combined branch of this and #41626, and then run the integration tests in a docker container for some older OSes. That should allow to figure this out (or at least allow to try some older versions).

felixfontein committed:
1.5 seems to work as well (earlier versions don't have EC sign function). Making Python 2.x adjustments.

felixfontein committed:
Changing option to select_crypto_backend.

felixfontein commented:
I settled on `select_crypto_backend` instead of `force_crypto_backend`.

felixfontein committed:
Python 2.6 compatibility.

resmo commented:
Pretty awesome work! I didn't see it but just for clarification: `select_crypto_backend: auto` would not break existing playbooks, right? Otherwise we have to document it in porting guide.

felixfontein commented:
Thanks!
Well, actually, that's a good question ;-) In theory, it should not break anything, since both methods should do the same computations. But of course it could happen that due to some badly set-up Python or some bad installation of the `cryptography` library, or due to something strange in a version I didn't explicitly test it with, it might not work. With that regard, `select_crypto_backend: openssl` would be the totally safe choice.
I'd still go for `select_crypto_backend: auto`, since it should work everywhere and generally should behave better (less things to go wrong, since there's no need to call `openssl` all the time; in particular, there's no need to exchange private stuff between the Python process and the `openssl` process).
How about using `auto` for now, and then decide a couple of weeks before the 2.7 merge window closes whether `openssl` is the better choice?

felixfontein commented:
In any case, I think it would be good to mention this in the porting guide, so that in case something goes wrong people find out faster what the problem is resp. how they can "fix" it.

felixfontein commented:
I just renewed all my certificate with this branch. :tada:

felixfontein commented:
Rebased to resolve conflicts with #42496.

felixfontein committed:
Trying to test both backends separately for acme_account.

felixfontein commented:
@mattclay I'd like your opinion on a test-related matter. This PR adds the possibility to choose between two backends for the `acme_*` modules. I'd like the tests to be run each with both backends.

For the `acme_account` module, I tried this out in ca69d0fe44363a5e395bd31943745880c71aa42f: I copied the `acme_account/tasks/main.yml` to `acme_account/tasks/impl.yml` and changed acme_account/tasks/main.yml to include the `impl.yml` twice, with different variables (which will select the different backends). I changed each `register: xxx` to `register: "{{ test_prefix }}xxx"`.
For `acme_account/tests/validate.yml`, I proceeded similarly, and in `acme_account/tests/validate_impl.yml` I changed things like `account_deactivate_idempotent.account_uri is not none` to `hostvars[test_prefix ~ 'account_deactivate_idempotent'].account_uri is not none`.
This looks a bit ugly, but works.

Do you think this is OK? Or do you have a better suggestion on how to run the tests with both backends without duplicating all the tests?

mattclay commented:
@felixfontein The first thing I noticed is that there doesn't appear to be anything actually including/importing `tests/validate.yml` in any of the acme_* integration tests.

As for actually running against two different backends, why not run the tasks and validation for the first backend, then remove the output directory and run the tests and validation for the next backend. That will avoid the need to have different filenames and variable names for each backend.

felixfontein commented:
@mattclay You're right! I saw this pattern in some other tests and copied it, but apparently overlooked the `- import_tasks: ../tests/validate.yml` at the end of each test... I assumed there's some kind of "magic" which runs the tests...
I guess I have to fix the tests first; then I'll rebase this PR and properly implement the tests as you suggest.

felixfontein committed:
Also testing both backends separately for acme_certificate and acme_certificate_revoke.

felixfontein committed:
Adding changelog entry which informs about select_crypto_backend option in case autodetect fails.

felixfontein commented:
@resmo I added a changelog entry in 7b551366437e40c45134df6ec547fc984753de6b which describes the new backend choice, and gives a hint on what to do if autodetect fails. Do you think this is sufficient (and understandable)?

felixfontein committed:
Fixing YAML.

felixfontein commented:
I've rebased to remove the conflicts with merged #42158.

felixfontein commented:
@resmo How should we proceed with this one?

felixfontein commented:
@resmo Thanks! That will also make the changes regarding ansible/acme-test-container#5 simpler, since there will be no need for a lot of conflict resolving ;)

felixfontein commented:
@resmo I think your super-powers are needed ;-)

resmo commented:
bot_status

felixfontein commented:
@resmo Thanks for merging! Looks like some files are missing the correct maintainer information. I'll create a PR for that, or at least I'll try ;)
